src/box/memtx_tx.c

```diff
@@ -2124,30 +2121,74 @@ memtx_tx_history_remove_story_del_stmts(struct memtx_story *story)
 static void
 memtx_tx_history_rollback_added_story(struct txn_stmt *stmt)
 {
-	assert(stmt->add_story != NULL);
-	assert(stmt->add_story->tuple == stmt->rollback_info.new_tuple);
-	struct memtx_story *story = stmt->add_story;
+	struct memtx_story *add_story = stmt->add_story;
+	struct memtx_story *del_story = stmt->del_story;
 
-	memtx_tx_history_remove_story_del_stmts(story);
-	story->add_stmt = NULL;
-	stmt->add_story = NULL;
+	/*
+	 * In case of rollback of prepared statement we need to rollback
+	 * preparation actions.
+	 */
+	if (stmt->txn->psn != 0) {
+		/*
+		 * During preparation of this statement there were two cases:
+		 * * del_story != NULL: all inprogress transactions that were
+		 *   to delete del_story were relinked to delete add_story.
+		 * * del_story == NULL: all inprogress transactions that were
+		 *   to delete same nothing were relinked to delete add_story.
+		 * See memtx_tx_history_prepare_insert_stmt for details.
+		 * Note that by design of rollback, all statements are rolbacked
+		 * in reverse order, and thus at this point there can be no
+		 * statements of the same transaction that deletes add_story.
+		 * So we must scan delete statements and relink them to delete
+		 * del_story if it's not NULL or to delete nothing otherwise.
+		 */
+		struct txn_stmt **from = &add_story->del_stmt;
+		while (*from != NULL) {
+			struct txn_stmt *test_stmt = *from;
+			assert(test_stmt->del_story == add_story);
+			assert(test_stmt->txn != stmt->txn);
+			assert(!test_stmt->is_own_change);
+			assert(test_stmt->txn->psn == 0);
+
+			/* Unlink from add_story list. */
+			*from = test_stmt->next_in_del_list;
+			test_stmt->next_in_del_list = NULL;
+			test_stmt->del_story = NULL;
+
+			if (del_story != NULL) {
+				/* Link to del_story's list. */
+				memtx_tx_story_link_deleted_by(del_story,
+							       test_stmt);
+			}
+		}
```

Comment:
In `READ_COMMITTED` mode `add_story->del_stmt` can be a statement of another in-progress transaction. It is not critical since we will abort such transactions after the next commit. Just write about it in comment (like you did in `rollback_deleted_story`), or you can try to fix this problem.

Reply:
Actually here `add_story->del_stmt` can be only statements of in-progress transactions, despite of isolation level. And there's not problem here: if I understand correctly in case of `memtx_tx_history_rollback_added_story` we can definitely revert all changes were done in `memtx_tx_history_prepare_added_story`.
Note that in the next commit we abort only reading transactions.